### PR TITLE
[serve] Simplify application build codepath

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Stacked on: https://github.com/ray-project/ray/pull/48223

Reimplements the `.bind()` codepath to avoid using the Ray DAG codepath which adds a lot of complexity for little benefit.

The DAG traversal code is much simpler, around 100 lines of self-contained code in `build_app.py`. I've also added unit tests for it.

There should be no behavior change here.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
NOTE: `test_deploy_2.py::test_deployment_error_handling` was failing with my changes because it was not actually checking the right thing previously. It was passing because a validation error was raised in the DAG building codepath, not in the controller. This meant that if bad actor options _did_ make it to the controller, it was not handled gracefully (similar to https://github.com/ray-project/ray/issues/43888).

I've updated the logic in `deployment_scheduler.py` to handle the case where Ray complains during actor creation and verified that it's handled gracefully now.

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/_private/build_app.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from ray.serve.handle import DeploymentHandle, _HandleOptions
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class IDDict(dict):
+    """Dictionary that uses id() for keys instead of hash().
+
+    This is necessary because Application objects aren't hashable and we want each
+    instance to map to a unique key.
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/build_app.py:17]: this will never be used across nodes

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/build_app.py:17]: but I agree that would probably be a safer solution overall

**Comment by edoakes** [comment]
now blocked on: https://github.com/ray-project/ray/pull/48294

**Pull request merged**